File: translated/f-to-c/DRB097-target-teams-distribute-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//use of omp target + teams + distribute + parallel for. No data race pairs.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main()
{
    long long i, i2, len, l_limit, tmp;
    double sum, sum2;
    double* a;
    double* b;

    len = 2560;
    sum = 0.0;
    sum2 = 0.0;

    a = (double*)malloc(len * sizeof(double));
    b = (double*)malloc(len * sizeof(double));

    for (i = 0; i < len; i++) {
        a[i] = (double)(i + 1) / 2.0;  // Adjust for 0-based indexing
        b[i] = (double)(i + 1) / 3.0;
    }

    #pragma omp target map(to: a[0:len], b[0:len]) map(tofrom: sum)
    #pragma omp teams num_teams(10) thread_limit(256) reduction(+:sum)
    #pragma omp distribute
    for (i2 = 0; i2 < len; i2 += 256) {
        #pragma omp parallel for reduction(+:sum)
        for (i = i2 + 1; i <= (i2 + 256 < len ? i2 + 256 : len); i++) {  // Adjust bounds
            sum = sum + a[i-1] * b[i-1];  // Adjust for 0-based indexing
        }
    }

    #pragma omp parallel for reduction(+:sum2)
    for (i = 0; i < len; i++) {
        sum2 = sum2 + a[i] * b[i];
    }

    printf("sum = %d; sum2 = %d\n", (int)sum, (int)sum2);

    free(a);
    free(b);

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB097-target-teams-distribute-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//use of omp target + teams + distribute + parallel for. No data race pairs.

package main

import (
	"fmt"
	"runtime"
	"sync"
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	var i, i2, length, lLimit, tmp int64
	var sum, sum2 float64
	var a, b []float64

	length = 2560
	sum = 0.0
	sum2 = 0.0

	a = make([]float64, length)
	b = make([]float64, length)

	for i = 1; i <= length; i++ {
		a[i-1] = float64(i) / 2.0
		b[i-1] = float64(i) / 3.0
	}

	//$omp target map(to: a(0:len), b(0:len)) map(tofrom: sum)
	//$omp teams num_teams(10) thread_limit(256) reduction (+:sum)
	var mu sync.Mutex
	//$omp distribute
	for i2 = 1; i2 <= length; i2 += 256 {
		//$omp parallel do reduction (+:sum)
		var wg sync.WaitGroup
		numCPU := runtime.NumCPU()
		end := min(i2+256, length)
		chunkSize := (end - i2) / numCPU
		if chunkSize < 1 {
			chunkSize = 1
		}

		for start := i2 + 1; start <= end; start += chunkSize {
			chunkEnd := start + chunkSize - 1
			if chunkEnd > end {
				chunkEnd = end
			}
			wg.Add(1)
			go func(start, end int64) {
				defer wg.Done()
				localSum := 0.0
				for i := start; i <= end; i++ {
					localSum = localSum + a[i-1]*b[i-1]
				}
				mu.Lock()
				sum = sum + localSum
				mu.Unlock()
			}(start, chunkEnd)
		}
		wg.Wait()
		//$omp end parallel do
	}
	//$omp end distribute
	//$omp end teams
	//$omp end target

	//$omp parallel do reduction (+:sum2)
	var wg2 sync.WaitGroup
	var mu2 sync.Mutex
	numCPU := runtime.NumCPU()
	chunkSize := length / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= length; start += chunkSize {
		end := start + chunkSize - 1
		if end > length {
			end = length
		}
		wg2.Add(1)
		go func(start, end int64) {
			defer wg2.Done()
			localSum := 0.0
			for i := start; i <= end; i++ {
				localSum = localSum + a[i-1]*b[i-1]
			}
			mu2.Lock()
			sum2 = sum2 + localSum
			mu2.Unlock()
		}(start, end)
	}
	wg2.Wait()
	//$omp end parallel do

	fmt.Printf("sum = %d; sum2 = %d\n", int(sum), int(sum2))
}
